source/common/protobuf/utility.cc

```diff
@@ -596,6 +599,118 @@ std::string MessageUtil::CodeEnumToString(ProtobufUtil::error::Code code) {
   }
 }
 
+namespace {
+
+std::unique_ptr<Protobuf::Message> typeUrlToNewMessage(Protobuf::MessageFactory& message_factory,
+                                                       absl::string_view type_url) {
+  const absl::string_view type_name = TypeUtil::typeUrlToDescriptorFullName(type_url);
+  const Protobuf::Descriptor* descriptor =
+      Protobuf::DescriptorPool::generated_pool()->FindMessageTypeByName(
+          static_cast<std::string>(type_name));
+  return std::unique_ptr<Protobuf::Message>(
+      descriptor == nullptr ? nullptr : message_factory.GetPrototype(descriptor)->New());
+}
+
+// Forward declaration for mutually-recursive helper functions.
+void redact(Protobuf::Message* message, bool ancestor_is_sensitive);
+
+// To redact an `Any`, we have to unpack it to its original type to redact it.
+bool redactAny(Protobuf::Message* message, bool ancestor_is_sensitive) {
+  auto* any = dynamic_cast<ProtobufWkt::Any*>(message);
+  if (any != nullptr) {
+    Protobuf::DynamicMessageFactory message_factory;
+    auto typed_message = typeUrlToNewMessage(message_factory, any->type_url());
+    if (typed_message == nullptr) {
+      // If the type URL doesn't correspond to a known proto, give up redacting and treat this
+      // message the same as a `ProtobufWkt::Struct`. See the documented limitation on
+      // `MessageUtil::redact()` for more context.
+      ENVOY_LOG_MISC(warn, "Could not redact ProtobufWkt::Any with unknown type URL {}",
+                     any->type_url());
+      return false;
+    }
+
+    any->UnpackTo(typed_message.get());
+    redact(typed_message.get(), ancestor_is_sensitive);
+    any->PackFrom(*typed_message);
+    return true;
+  }
+  return false;
+}
+
+// To redact a `TypedStruct`, we have to reify it based on its `type_url` to redact it.
+bool redactTypedStruct(Protobuf::Message* message, bool ancestor_is_sensitive) {
+  auto* typed_struct = dynamic_cast<udpa::type::v1::TypedStruct*>(message);
```

Comment:
I'm not sure how safe this is in general. You can have a `TypeStruct` message created by a `DynamicMessageFactory`, which isn't C++ dynamic castable. We are doing this trick increasingly in the `VersionConverter` (I have two active PRs that do this). I'm wondering if we should check the proto descriptor's `full_name` here, and use reflection to extract out the fields.

Reply:
You are right, they are of type `google::protobuf::DynamicMessage`. I will update.

Reply:
Done, this was a really important catch. PTAL, thanks!